pageserver/src/tenant/mgr.rs

```diff
@@ -524,30 +562,60 @@ pub async fn attach_tenant(
     remote_storage: GenericRemoteStorage,
     ctx: &RequestContext,
 ) -> Result<(), TenantMapInsertError> {
-    tenant_map_insert(tenant_id, || {
-        let tenant_dir = create_tenant_files(conf, tenant_conf, tenant_id, CreateTenantFilesMode::Attach)?;
-        // TODO: tenant directory remains on disk if we bail out from here on.
-        //       See https://github.com/neondatabase/neon/issues/4233
+    let func = || async {
+        let tenant_dir =
+            create_tenant_files(conf, tenant_conf, tenant_id, CreateTenantFilesMode::Attach)?;
+
+        let guard_fs = scopeguard::guard((), |_| {
+            if let Err(e) = std::fs::remove_dir_all(&tenant_dir) {
+                // log the error but not throwing it somewhere
+                warn!("failed to cleanup when tenant {tenant_id} fails to start: {e}");
+            }
+        });
 
         // Without the attach marker, schedule_local_tenant_processing will treat the attached tenant as fully attached
         let marker_file_exists = conf
             .tenant_attaching_mark_file_path(&tenant_id)
             .try_exists()
             .context("check for attach marker file existence")?;
-        anyhow::ensure!(marker_file_exists, "create_tenant_files should have created the attach marker file");
+        anyhow::ensure!(
+            marker_file_exists,
+            "create_tenant_files should have created the attach marker file"
+        );
 
-        let attached_tenant = schedule_local_tenant_processing(conf, &tenant_dir, broker_client, Some(remote_storage), None, ctx)?;
+        let attached_tenant = schedule_local_tenant_processing(
+            conf,
+            &tenant_dir,
+            broker_client,
+            Some(remote_storage),
+            None,
+            ctx,
+        )?;
         // TODO: tenant object & its background loops remain, untracked in tenant map, if we fail here.
         //      See https://github.com/neondatabase/neon/issues/4233
 
-        let attached_tenant_id = attached_tenant.tenant_id();
-        anyhow::ensure!(
-            tenant_id == attached_tenant_id,
-            "loaded created tenant has unexpected tenant id (expect {tenant_id} != actual {attached_tenant_id})",
-        );
+        // Put all code that might error in the check function.
+
+        let check = || {
+            let attached_tenant_id = attached_tenant.tenant_id();
+            anyhow::ensure!(
+                tenant_id == attached_tenant_id,
+                "loaded created tenant has unexpected tenant id (expect {tenant_id} != actual {attached_tenant_id})",
+            );
+            Ok(())
+        };
+
+        if let Err(e) = check() {
+            attached_tenant.set_broken("failed to create".into()).await;
+            return Err(e);
+        }
+
```

Comment:
The missing comment here is:

```rust
// schedule_local_tenant_processing eventually launches the tenant's background task
// We need to shut them down before bailing out.
```

Note further that `set_broken` doesn't necessarily stop all background activity. `Tenant::shutdown` is the new way to do it.

When addressing this, also remove the TODO comment a few lines above.

cc @koivunej, might be interesting to look at this use case for next steps in shutdown work.

Reply:
Hmm this PR seems to have changed quite a lot since I approved it, trying to understand what all has been rewritten. Changes do look quite odd at first glance.

Reply:
It would probably make sense if the load/create/attach time spawned task would be the sole task for the tenant.

Reply:
addressed in https://github.com/neondatabase/neon/pull/4257/commits/5f8bd417259cce7ef8f97e06297edb17086f23c3